cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll

```diff
@@ -257,6 +257,17 @@ private predicate needsLoadForParentExpr(Expr expr) {
   exists(CrementOperation crement | expr = crement.getOperand().getFullyConverted())
   or
   exists(AssignOperation ao | expr = ao.getLValue().getFullyConverted())
+  or
+  // For arguments that are passed by value but require a constructor call, the extractor emits a
+  // `TemporaryObjectExpr` as the argument, and marks it as a glvalue. This is roughly how a code-
+  // generating compiler would implement this, passing the address of the temporary so that the
+  // callee is using the exact same memory location allocated by the caller. We don't fully model
+  // this yet, though, so we'll synthesize a load so that we appear to be passing the temporary
+  // object via a bitwise copy.
+  exists(Call call |
+    expr = call.getAnArgument().getFullyConverted().(TemporaryObjectExpr) and
+    expr.isGLValueCategory()
+  )
```

Comment:
I think we need to exclude the qualifier here, and instead use the `VariableAddressInstruction` generated for the `TemporaryObjectExpr` as the `this` operand for the call. See the comment in the tests.

Reply:
Ah, good catch, although the bug isn't in this predicate. The issue is that C++ 98 allowed the qualifier of a member access to be an rvalue, so EDG marks the temporary object created as the qualifier to be a prvalue. More recent standards changed the wording to that if the qualifier is a prvalue, it undergoes temporary object materialization to turn it into a glvalue, so that the converted qualifier is always a glvalue. EDG presumably preserves the load for compatibility, but we should ignore it. This predicate won't re-add the load, because `call.getAnArgument()` does not include the qualifier.

The same problem exists with field accesses on prvalues. I'll have a fix and updated tests soon.

Reply:
Fixed.